### PR TITLE
[JUJU-1925] Secret get using label

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR adds support for `secret-get` hook command to fetch a secret using either `secret label` for owners or `consumer label` for consumers. 

Drive-by: set the MinMajorUpgradeVersion to `2.9.36` for the juju3 upgrade.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```console
$ juju exec --unit hello-kubecon/0 'uri=$(secret-add foo=app); echo $uri; secret-set $uri --label=app;'
secret:cd3s6aeffbal34dgcs60

$ juju exec --unit hello-kubecon/0 "secret-get cd3s6aeffbal34dgcs60"
foo: app

$ juju exec --unit hello-kubecon/0 "secret-get --label app"
ERROR consumer label "app" not found
ERROR the following task failed:
 - id "8" with return code 1

use 'juju show-task' to inspect the failure

$ juju exec --unit hello-kubecon/0 "secret-get cd3s6aeffbal34dgcs60 --label consumer-app"
foo: app

$ juju exec --unit hello-kubecon/0 "secret-get --label consumer-app"
foo: app

$ juju exec --unit hello-kubecon/0 'uri=$(secret-add --owner unit foo=unit0); echo $uri; secret-set $uri --label=unit;'
secret:cd3s65uffbal34dgcs5g

$ juju exec --unit hello-kubecon/0 "secret-get secret:cd3s65uffbal34dgcs5g"
foo: unit0

$ juju exec --unit hello-kubecon/0 "secret-get --label unit"
foo: unit0

$ juju exec --unit hello-kubecon/0 "secret-get cd3s65uffbal34dgcs5g --label unit --update"
ERROR secret owner cannot use --update
ERROR the following task failed:
 - id "28" with return code 1

use 'juju show-task' to inspect the failure

$ juju exec --unit hello-kubecon/0 "secret-grant -r 0 secret:cd3s65uffbal34dgcs5g"

$ juju exec --unit nginx-ingress-integrator/0 "secret-get cd3s65uffbal34dgcs5g"
foo: unit0

$ juju exec --unit nginx-ingress-integrator/0 "secret-get --label unit"
ERROR consumer label "unit" not found
ERROR the following task failed:
 - id "34" with return code 1

use 'juju show-task' to inspect the failure

$ juju exec --unit nginx-ingress-integrator/0 "secret-get --label nginx-ingress-integrator"
ERROR consumer label "nginx-ingress-integrator" not found
ERROR the following task failed:
 - id "36" with return code 1

use 'juju show-task' to inspect the failure

$ juju exec --unit nginx-ingress-integrator/0 "secret-get cd3s65uffbal34dgcs5g --label nginx-ingress-integrator"
foo: unit0

$ juju exec --unit nginx-ingress-integrator/0 "secret-get --label nginx-ingress-integrator"
foo: unit0

```

## Documentation changes

TBD

## Bug reference

No

### CODE REVIEW

File: state/application.go [hunk 1]
```diff
@@ -694,6 +694,12 @@ func (a *Application) removeOps(asserts bson.D, op *ForcedOperation) ([]txn.Op,
 	}
 	ops = append(ops, secretLabelOps...)
 
+	secretLabelOps, err = a.st.removeConsumerSecretLabelOps(a.ApplicationTag())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	ops = append(ops, secretLabelOps...)
+
```
Comment (wallyworld) [inline state/application.go:702]: This isn't needed - consumers are not apps but individual units

File: state/secrets.go [hunk 2]
```diff
@@ -646,19 +642,37 @@ func (s *secretsStore) getSecretValue(uri *secrets.URI, revision int, checkExist
 }
 
 // GetSecret gets the secret metadata for the specified URL.
```
Comment (wallyworld) [inline state/secrets.go:644]: Update comment

File: state/secrets.go [hunk 3]
```diff
@@ -877,21 +899,32 @@ func (st *State) checkConsumerCountOps(uri *secrets.URI, inc int) ([]txn.Op, err
...
-	key := secretOwnerLabelKey(ownerTag, label)
+	key := keyGenerator(tag, label)
 	countOp, count, err := nsRefcounts.CurrentOp(refCountCollection, key)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	if count > 0 {
-		return nil, errors.WithType(
-			errors.Errorf("secret label %q owned by %s already exists", label, ownerTag), LabelExists)
+		return nil, errors.WithType(errors.Errorf("label %q for %q already exists", label, tag), LabelExists)
```
Comment (wallyworld) [inline state/secrets.go:927]: we should retain "secret" in the message. also, can we add "owner" or "consumer", ie pass in a badge param
`errors.Errorf("secret %s label %q for %q already exists", badge, label, tag)`

File: worker/uniter/runner/jujuc/secret-get.go [hunk 4]
```diff
@@ -91,21 +91,24 @@ func (c *secretGetCommand) Init(args []string) (err error) {
...
+	if c.secretUri == nil && c.label == "" {
+		return errors.New("require either a secret URI or label")
+	}
+
+	if !c.update {
 		if c.secretUri != nil && c.label != "" {
-			return errors.New("specify either a secret URI or label but not both to fetch metadata")
+			return errors.New("specify either a secret URI or label but not both")
 		}
+	}
```
Comment (wallyworld) [inline worker/uniter/runner/jujuc/secret-get.go:103]: this goes below the metadata check

File: api/agent/secretsmanager/client.go [hunk 5]
```diff
@@ -78,21 +78,23 @@ func (c *Client) GetContentInfo(uri *coresecrets.URI, label string, update, peek
 		Update: update,
 		Peek:   peek,
 	}
-	arg.URI = uri.String()
+	if uri != nil {
```
Comment (wallyworld) [inline api/agent/secretsmanager/client.go:81]: probably want to add a new test for this scenario

**Review by wallyworld — COMMENTED** [review]
Some initial comments

File: state/allcollections.go [hunk 6]
```diff
@@ -542,7 +542,7 @@ func allCollections() CollectionSchema {
 
 		secretMetadataC: {
 			indexes: []mgo.Index{{
-				Key: []string{"model-uuid", "_id"},
+				Key: []string{"owner-tag", "label", "model-uuid"},
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply state/allcollections.go:545]: `_id` is indexed by default.

File: state/application.go (hunk 1 above)
Comment (ycliuhw) [reply state/application.go:702]: operator consumers are all apps?

File: worker/uniter/runner/jujuc/secret-get.go (hunk 4 above)
Comment (ycliuhw) [reply worker/uniter/runner/jujuc/secret-get.go:103]: We should only allow `URI` + `--label` for `--update`, right?

Comment (ycliuhw) [reply worker/uniter/runner/jujuc/secret-get.go:103]: So after discussed, we allow `secret-get $URI --label label1` to create a consumer label.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 7]
```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
...
+	if s.authTag.String() != md.OwnerTag {
+		// Invisible for non owner users.
+		return nil, errors.NotFoundf("secret %s, label %q", uri, label)
+	}
+	return md, nil
+}
+
+func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
+	// Only the owner can access secrets via the secret label (Note: the leader unit is not the owner of the application secrets).
+	// Non owners have to use secret URI or consumer label.
```
Comment (wallyworld) [inline apiserver/facades/agent/secretsmanager/secrets.go:369]: // Only the owner can access secrets via the secret metadata label added by the owner.
// (Note: the leader unit is not the owner of the application secrets).
// Consumers get to use their own label.
// Both owners and consumers can also just use the secret URI.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 8]
```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
...
+	var err error
+
+	if arg.URI != "" {
+		uri, err = coresecrets.ParseURI(arg.URI)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	// arg.Label is the secret label for owner.
```
Comment (wallyworld) [inline apiserver/facades/agent/secretsmanager/secrets.go:397]: arg.Label could also be the consumer label too
The comment here should be along the lines of we will first check to see if the label corresponds to a secret owned by the caller.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 9]
```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
...
+	md, err := s.getOwnerSecretMetadata(uri, arg.Label)
+	if err != nil && !errors.Is(err, errors.NotFound) {
+		return nil, errors.Trace(err)
+	}
+	if md != nil {
+		// Owner can access the content directly.
+		if arg.Update {
+			// We should not have cmd flag knowledge here but we have to check here because jujuc cannot know it.
+			return nil, errors.NewNotValid(nil, "secret owner cannot use --update")
+		}
```
Comment (wallyworld) [inline apiserver/facades/agent/secretsmanager/secrets.go:407]: We need to check this in the hook context before making this api call. The hook context has pre-loaded all the metadata fore secrets owned by the caller. So we can look up the owned secret URI in the hook context and just pass the URI to the GetSecretValue facade call.

See contextfactory.go
```
	info, err := ctx.secretsClient.SecretMetadata(secrets.Filter{
		OwnerTag: &owner,
	})
```

The only time we will not have a uri passed in is if a secret consumer is making the api call. In this case, we do not even look up any owner metadata - we just search for consumer metadata.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 10]
```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
...
+	if uri == nil {
+		var err error
+		uri, err = s.secretsConsumer.GetURIByConsumerLabel(arg.Label, s.authTag)
+		if errors.Is(err, errors.NotFound) {
+			return nil, errors.NotFoundf("consumer label %q", arg.Label)
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
```
Comment (wallyworld) [inline apiserver/facades/agent/secretsmanager/secrets.go:423]: This is the only URI look up we need to do - owners will always pass in a URI. If the owner attempts to use --label and also gives a URI, that's an error.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 11]
```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
 	return result, nil
 }
 
-func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
-	uri, err := coresecrets.ParseURI(arg.URI)
+func (s *SecretsManagerAPI) getOwnerSecretMetadata(uri *coresecrets.URI, label string) (*coresecrets.SecretMetadata, error) {
```
Comment (wallyworld) [inline apiserver/facades/agent/secretsmanager/secrets.go:347]: This isn't needed as we can look up what we need on the hook context

**Review by wallyworld — APPROVED** [review]

**Comment by hpidcock** [comment]
https://bugs.launchpad.net/juju/+bug/1991806

**Pull request merged**